### PR TITLE
Aurora database integration

### PR DESCRIPTION
**Pull request by parthshahva**
Things left
1) Manual testing of validation and checksummer images
2) Review code climate
3) get tests to pass in travis
4) Edgcase and exception handling + logging
5) update readme with new dev and deployment instructions

Story of commits:
1) Terraform setup of database module 
2) Alembic integration for management of schema and migrations
3) Database.py module and tests using sqlalchemy library (library that alembic is built upon as well)
4) DB interface for upload area, uploaded file, and ingest notifier
5) changes/additions to upload api (2 new endpoints) and inclusion of ingest notifier from upload api with tests
6) client method for updating validation and checksum events called from docker images with tests
7) changes to checksumming flow to use new data flow and classes, no more immediate checksumming
8) changes to validation flow to use new data flow and classes

### CODE REVIEW

File: docker-images/base-alpine-python36/harness.py [hunk 1]
```diff
@@ -30,23 +30,51 @@ def get_logger(name, corr_id_name, corr_id_val):
...
+        validation_event = UploadedFileValidationEvent(file_id=self.s3_object_key,
+                                                       validation_id=self.id,
+                                                       job_id=self.validation_id,
+                                                       status="VALIDATING")
+        if not args.test:
+            update_event(validation_event)
         results = self._run_validator()
         self._unstage_file()
-        self._report_results(results)
+        validation_event.status = "VALIDATED"
```
Comment (parthshahva) [inline docker-images/base-alpine-python36/harness.py]: change this to COMPLETED( can also have a failure state) and also store results in a json b column

File: tests/lambdas/api_server/test_area.py [hunk 2]
```diff
@@ -1,15 +1,20 @@
 #!/usr/bin/env python3.6
 
 import os, sys, unittest, uuid, json, base64
-
```
Comment (sampierson) [inline tests/lambdas/api_server/test_area.py]: Please leave blank line between stdlib, 3rd party, and our code, imports.

File: tests/lambdas/api_server/test_area.py [hunk 3]
```diff
@@ -1,15 +1,20 @@
...
 import boto3
 from botocore.exceptions import ClientError
 from moto import mock_s3, mock_iam, mock_sns, mock_sts
-from unittest.mock import patch
-
+from unittest.mock import patch, Mock
```
Comment (sampierson) [inline tests/lambdas/api_server/test_area.py]: ditto

File: upload/common/ingest_notifier.py [hunk 4]
```diff
@@ -17,8 +17,18 @@ class IngestNotifier:
     FILE_UPLOAD_EXCHANGE = 'ingest.file.staged.exchange'
     FILE_UPLOADED_QUEUE = 'ingest.file.create.staged'
 
-    def __init__(self):
-        self.id = str(uuid.uuid4())
+    FILE_VALIDATED_EXCHANGE = "ingest.validation.exchange"
```
Comment (sampierson) [inline upload/common/ingest_notifier.py]: Did you clear this exchange change with the Ingest team?  They gave me this endpoint.

File: upload/common/ingest_notifier.py [hunk 5]
```diff
@@ -17,8 +17,18 @@ class IngestNotifier:
...
+    FILE_VALIDATED_QUEUE = "ingest.file.validation.queue"
```
Comment (sampierson) [inline upload/common/ingest_notifier.py]: ditto

File: upload/lambdas/api_server/v1/area.py [hunk 6]
```diff
@@ -1,16 +1,19 @@
 import json
 import urllib.parse
-
```
Comment (sampierson) [inline upload/lambdas/api_server/v1/area.py]: blank line

File: upload/lambdas/api_server/v1/area.py [hunk 7]
```diff
@@ -1,16 +1,19 @@
...
 import connexion
 import requests
-
```
Comment (sampierson) [inline upload/lambdas/api_server/v1/area.py]: ditto

File: upload/common/upload_api_client.py [hunk 8]
```diff
@@ -0,0 +1,30 @@
+from urllib.parse import urljoin
```
Comment (sampierson) [inline upload/common/upload_api_client.py]: Is this really "common" code or is the api_client worthy of its own folder under upload/?  Not sure.

File: tests/common/test_checksum.py [hunk 9]
```diff
@@ -1,8 +1,8 @@
 import unittest
 import uuid
-
```
Comment (sampierson) [inline tests/common/test_checksum.py]: blank

File: tests/lambdas/test_checksum_daemon.py [hunk 10]
```diff
@@ -76,46 +84,12 @@ def tearDown(self):
...
-            'url': f"s3://{self.UPLOAD_BUCKET_NAME}/{self.area_id}/{self.filename}",
-            'checksums': FIXTURE_DATA_CHECKSUMS[self.file_contents]['checksums']
-        })
-
     @patch('upload.common.upload_area.UploadedFile.size', 100 * 1024 * 1024 * 1024)
     @patch('upload.lambdas.checksum_daemon.checksum_daemon.ChecksumDaemon.schedule_checksumming')
     @patch('upload.lambdas.checksum_daemon.checksum_daemon.IngestNotifier.connect')
-    @patch('upload.lambdas.checksum_daemon.checksum_daemon.IngestNotifier.file_was_uploaded')
+    @patch('upload.lambdas.checksum_daemon.checksum_daemon.IngestNotifier.format_and_send_notification')
     def test_that_with_a_large_file_a_batch_job_is_scheduled(self,
```
Comment (sampierson) [inline tests/lambdas/test_checksum_daemon.py]: Maybe rename to "test_that_a_batch_job_is_scheduled"

File: upload/common/checksum.py [hunk 11]
```diff
@@ -1,8 +1,8 @@
 import time
-
```
Comment (sampierson) [inline upload/common/checksum.py]: etc

File: upload/common/checksum.py [hunk 12]
```diff
@@ -37,7 +37,8 @@ def checksum(self, report_progress=False):
         else:
             progress_callback = None
 
-        return self._compute_checksums(progress_callback=progress_callback)
+        self.checksums = self._compute_checksums(progress_callback=progress_callback)
```
Comment (sampierson) [inline upload/common/checksum.py]: I don't see this instance variable used anywhere.  Not sure there's real value in assigning to a variable before immediately returning it.  If you're trying to document what the function is returning, a local variable would be fine for that.

File: upload/common/checksum_event.py [hunk 13]
```diff
@@ -0,0 +1,39 @@
+from datetime import datetime
```
Comment (sampierson) [inline upload/common/checksum_event.py]: I wonder if a folder for events might be appropriate.

File: upload/docker_images/checksummer/checksummer.py [hunk 14]
```diff
@@ -3,14 +3,14 @@
 import argparse
 import os
 import sys
-
```
Comment (sampierson) [inline upload/docker_images/checksummer/checksummer.py]: Please follow the conventions of the codebase you're working in, or ask.  I believe the data store also uses this convention.  That's where I got it, and it makes sense.

File: database/versions/3b92db7bb2fe_create_initial_tables.py [hunk 15]
```diff
@@ -0,0 +1,114 @@
+"""create initial tables
```
[the remark this answers is not in the document]
Comment (sampierson) [reply database/versions/3b92db7bb2fe_create_initial_tables.py]: Feel free to ignore this.  It's not THAT important, but it is a good pattern to follow for the future.

File: terraform/modules/database/database.tf [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+  program = ["python", "${path.root}/../../../scripts/process_db_secrets.py"]
+ 
+  query = {  
+    # arbitrary map from strings to strings, passed  
+    # to the external program as the data query. 
+    secret_string = "${data.aws_secretsmanager_secret_version.db_creds.secret_string}"
+ }
+}
+
+resource "aws_rds_cluster_instance" "cluster_instances" {
```
[the remark this answers is not in the document]
Comment (parthshahva) [reply terraform/modules/database/database.tf]: There are currently instances within a cluster yes. However, aurora postgresql serverless is coming this summer. We'll transition over as soon as this happens.

File: database/versions/3b92db7bb2fe_create_initial_tables.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (parthshahva) [reply database/versions/3b92db7bb2fe_create_initial_tables.py]: Ignoring for this PR but great pattern to use in the future!

**Pull request merged**